src/storage/free_space_manager.cpp

```diff
@@ -169,7 +169,15 @@ void FreeSpaceManager::mergePageRanges(free_list_t newInitialEntries) {
             prevEntry = entry;
         }
     }
-    addFreePages(prevEntry);
+    handleLastPageRange(prevEntry, fileHandle);
+}
+
+void FreeSpaceManager::handleLastPageRange(PageRange pageRange, FileHandle* fileHandle) {
+    if (pageRange.startPageIdx + pageRange.numPages == fileHandle->getNumPages()) {
+        fileHandle->removePageIdxAndTruncateIfNecessary(pageRange.startPageIdx);
```

Comment:
It's not really a problem with this code, but `removePageIdxAndTruncateIfNecessary` isn't actually truncating the file, just the FileHandle, so this could still lead to unused pages if the database is closed and re-loaded before the truncated pages are re-added.

Having it truncate the file causes `CopyTest.NodeInsertBMExceptionDuringCheckpointRecovery` to fail (I think it's trying to update a page past the end of the file when recovering).